src/createServer.js

```diff
@@ -1,11 +1,244 @@
+
+  let users = [];
+  let expenses = [];
+
+  app.get('/expenses', express.json(), (req, res) => {
+    const {
+      userId,
+      categories,
+      from,
+      to,
+    } = req.query;
+
+    const id = +userId;
+    const correctUser = users.find(user => user.id === userId);
+    const filteredExpenses = expenses.filter(expense =>
+      expense.userId === id);
+
+    if (!expenses.length) {
+      res.send([]);
+
+      return;
+    }
+
+    if (from && to) {
+      const expensesFromData = expenses.filter(
+        expense => expense.spentAt >= from && expense.spentAt <= to
+      );
+
+      res.statusCode = 200;
+      res.send(expensesFromData);
+
+      return;
+    }
+
+    if (correctUser && categories) {
+      const expensesCategory = filteredExpenses.filter(
+        expense => expense.category === categories
+      );
+
+      res.statusCode = 200;
+      res.send(expensesCategory);
+
+      return;
+    }
+
+    if (correctUser) {
+      res.statusCode = 200;
+      res.send(filteredExpenses);
+
+      return;
+    }
+
+    res.statusCode = 200;
+    res.send(expenses);
+  });
+
+  app.get('/expenses/:id', express.json(), (req, res) => {
+    const { id } = req.params;
+
+    const idExpense = Number(id);
+    const correctExpense = expenses.find(expense => expense.id === idExpense);
+
+    if (!correctExpense) {
+      res.sendStatus(404);
+
+      return;
+    }
+
+    res.statusCode = 200;
+
+    res.send(correctExpense);
+  });
+
+  app.post('/expenses', express.json(), (req, res) => {
+    const { userId } = req.body;
+    const newExpense = {
+      id: Math.floor(Math.random() * 10),
+      ...req.body,
+    };
+
+    if (!users.some(user => user.id === userId)) {
+      res.sendStatus(400);
+
+      return;
+    }
+
+    expenses.push(newExpense);
+
+    res.statusCode = 201;
+    res.send(newExpense);
+  });
+
+  app.delete('/expenses/:id', express.json(), (req, res) => {
+    const { id } = req.params;
+    const correctExpense = expenses.filter(expense => expense.id !== +id);
+
+    if (correctExpense.length === users.length) {
+      res.sendStatus(404);
+
+      return;
+    }
+
+    expenses = correctExpense;
+
+    res.sendStatus(204);
+  });
+
+  app.patch('/expenses/:id', express.json(), (req, res) => {
+    const { id } = req.params;
+    const correctExpense = expenses.find(expense => expense.id === +id);
+
+    if (typeof +id !== 'number') {
+      res.sendStatus(400);
+
+      return;
+    }
+
+    if (!correctExpense) {
+      res.sendStatus(404);
+
+      return;
+    }
+
+    const { title } = req.body;
+
+    if (typeof title !== 'string') {
+      res.sendStatus(404);
+
+      return;
+    }
+
+    Object.assign(correctExpense, { title });
+
+    res.statusCode = 200;
+
+    res.send(correctExpense);
+  });
+
+  app.get('/users', express.json(), (req, res) => {
+    if (!users.length) {
+      res.send([]);
+
+      return;
+    }
+    res.statusCode = 200;
+    res.send(users);
+  });
```

Comment:
do you need this conditional? if users.length is 0, users is []
```suggestion
  app.get('/users', express.json(), (req, res) => {
    res.statusCode = 200;
    res.send(users);
  });
```